File: JavaScript/Graph/L_allAncestorInDAG.js
var getAncestors = function (n, edges) {
    let res = Array.from({ length: n }, () => []);
    let graph = Array.from({ length: n }, () => []);

    for (let edge of edges) {
        graph[edge[0]].push(edge[1]);
    }

    for (let i = 0; i < n; i++) {
        dfs(graph, i, i, res, Array(n).fill(false));
    }

    for (let i = 0; i < n; i++) {
        res[i].sort((a, b) => a - b);
    }

    return res;
};

function dfs(graph, parent, curr, res, visit) {
    visit[curr] = true;
    for (let dest of graph[curr]) {
        if (!visit[dest]) {
            res[dest].push(parent);
            dfs(graph, parent, dest, res, visit);
        }
    }
}

File: Python/Graphs/L_allAncestorInDAG.py
def getAncestors(n, edges):
    res = [[] for _ in range(n)]
    graph = [[] for _ in range(n)]
    
    for edge in edges:
        graph[edge[0]].append(edge[1])
    
    for i in range(n):
        dfs(graph, i, i, res, [False] * n)
    
    for i in range(n):
        res[i].sort()
    
    return res
    
def dfs(graph, parent, curr, res, visit):
    visit[curr] = True
    for dest in graph[curr]:
        if not visit[dest]:
            res[dest].append(parent)
            dfs(graph, parent, dest, res, visit)
    